// ahooks/lib/useSet/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _react = require("react");
var _useMemoizedFn = _interopRequireDefault(require("../useMemoizedFn"));
function useSet(initialValue) {
  var getInitValue = function getInitValue() {
    return initialValue === undefined ? new Set() : new Set(initialValue);
  };
  var _useState = (0, _react.useState)(function () {
      return getInitValue();
    }),
    _useState2 = (0, _slicedToArray2["default"])(_useState, 2),
    set = _useState2[0],
    setSet = _useState2[1];
  var add = function add(key) {
    if (set.has(key)) {
      return;
    }
    setSet(function (prevSet) {
      var temp = new Set(prevSet);
      temp.add(key);
      return temp;
    });
  };
  var remove = function remove(key) {
    if (!set.has(key)) {
      return;
    }
    setSet(function (prevSet) {
      var temp = new Set(prevSet);
      temp["delete"](key);
      return temp;
    });
  };
  var reset = function reset() {
    return setSet(getInitValue());
  };
  return [set, {
    add: (0, _useMemoizedFn["default"])(add),
    remove: (0, _useMemoizedFn["default"])(remove),
    reset: (0, _useMemoizedFn["default"])(reset)
  }];
}
var _default = useSet;
exports["default"] = _default;
